### ISSUE TITLE
ERC721 Vyper contract bytes4 bug

### ISSUE THREAD

**Issue by liam-ot** [vyperlang/vyper#2945]
### Version Information

* vyper Version (output of `vyper --version`): 0.3.3
* OS: osx
* Python Version (output of `python --version`): Python 3.9.9

### What's your issue about?

The vyper language currently does not support the `bytes4` type which makes implementing the ERC165 standard in ERC721 contracts impossible.

The official Vyper ERC721 contract I am using is [here](https://github.com/vyperlang/vyper/blob/master/examples/tokens/ERC721.vy)

```
@external
def safeTransferFrom(
        _from: address,
        _to: address,
        _tokenId: uint256,
        _data: Bytes[1024]=b""
    ):
        …
        assert returnValue == method_id("onERC721Received(address,address,uint256,bytes)", output_type=bytes4)
```

The final line where the `assert` is called is where this error is occurring

The full error message is:

```
vyper.exceptions.ArgumentException: output_type must be bytes[4] or bytes32 vyper.exceptions.ArgumentException: output_type must be bytes[4] or bytes32
contract "nft-vy/contracts/nft.vy", function "safeTransferFrom", line 282:91
281 # https://github.com/eth-brownie/brownie/issues/1566
---> 282 assert returnValue == method_id("onERC721Received(address,address,uint256,bytes)", output_type=bytes4)
----------------------------------------------------------------------------------------------------^
283
```

an attempt to change to the recommended `bytes32` or `bytes[4]` also cause separate issues

@charles-cooper and I had discussed this shortly in #2619 and decided it was a good idea to move the discussion to a separate issue

### How can it be fixed?

Adding support for the `bytes4` type

**Comment by charles-cooper** [comment]
@liam-ot could you try adding a version pragma to the file?
```
# @version 0.3.3
```

**Comment by fubuloubu** [comment]
It supports the `bytes4` type, I think the problem is here is that the `method_id` function does not support outputting as `bytes4`

I believe we discussed this, but I can't see to find the linked issue, the resolution would be to change the `method_id` function to output `bytes4` as default (breaking change so in v0.4.0) and then allow selecting `bytes32` or `Bytes[4]` options as well (or deprecating them to force use of `convert` instead

**Comment by charles-cooper** [comment]
in v0.3.4 `method_id` will output `bytes4` or `Bytes[4]` -- https://github.com/vyperlang/vyper/commit/46ec8029815b1b59c35641c49924cd3dbb98f509

for v0.3.3, @liam-ot what happens when you try to output `Bytes[4]`?

**Comment by liam-ot** [comment]
@charles-cooper `Cannot perform equality between dislike types` because the assert compares the output of `method_id` and `returnValue: bytes4 = ERC721Receiver(_to).onERC721Received(msg.sender, _from, _tokenId, _data)` this variable

**Comment by liam-ot** [comment]
@fubuloubu i see now that it does, whoops mispoke when i said that then

**Comment by liam-ot** [comment]
@charles-cooper this compiles

```
assert returnValue == convert(method_id("onERC721Received(address,address,uint256,bytes)", output_type=bytes32), bytes4)
```

**Comment by charles-cooper** [comment]
@liam-ot is this still an issue?

**Comment by charles-cooper** [comment]
sounds good -- no rush!

**Comment by charles-cooper** [comment]
@liam-ot as of 0.3.4 `method_id()` allowable output should work for your use case. closing now for tidiness, please reopen if you still have an issue

**Comment by liam-ot** [comment]
this is fully resolved in 0.3.4 thanks very much!
